woocommerce/woocommerce
# Reduce code duplication in Report Controllers

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:
This code does not change any functionality. Just move code around to make the maintenance more manageable and highlight inconsistencies.

- [Remove `$rest_base` from `…\API\Reports\Controller`](https://github.com/woocommerce/woocommerce/commit/c924b21c844e52add22f293a38d63fd75f4d7f8f) 
   as it is defined in its parent `WC_REST_Controller`
- [Move `X-WP-Total` headers logic to a shared class,](https://github.com/woocommerce/woocommerce/commit/7a2a80bf8118b14935688dbb9c85a8d064915fab) 
- [Move common `get_collection_params` bits to the shared class](https://github.com/woocommerce/woocommerce/commit/9126db1abb3472df00eeb8d2489ed43afbf0b5d3) 
- [Move common `prepare_item_for_response` bits to the shared class](https://github.com/woocommerce/woocommerce/commit/9f921b8e404f8ff9dc323ecc674a4e1602fe0050) 
- [Move `interval` collection param to a shared parent class](https://github.com/woocommerce/woocommerce/pull/38048/commits/58bc88e0fcad82844182bd5f4fc12183fbc8f721)
- [Move common stats item schema bits to a shared class](https://github.com/woocommerce/woocommerce/pull/38048/commits/b0b80445aba7e476fac2229f722ed786ba6a49b2)

`Reports/Controller` differs in terms of `prepare_item_for_response` & `get_collection_params`, so its children will still contain redundant code.

We could consider improving it as well.

### How to test the changes in this Pull Request:

As mentioned above, this code does not change any functionality, so we could read the code to check if it is actually easier to maintain and check for regression issues.



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Read the code to check if it's shorter and easier to comprehend.
1. Run all automated tests
2. Smoke tests Analytics reports
